src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs

```diff
@@ -0,0 +1,1504 @@
+        Array.Copy(oary,nary,util.ShardSize)
+        nary
+
+    let mapList () =
+        let mutable result = []
+        for bi in 0 .. bucket.Length - 1 do
+            for si in 0 .. ShardSize - 1 do
+                if not(util. isNull bucket.[bi].[si]) then
+                    result <- bucket.[bi].[si] :: result
+        result                                 
+
+    let getMap (key:'K) =
+        let kh = key.GetHashCode()
+        bucket.[util.bucketIndex(kh,bucketBitMask)].[util.shardIndex kh]
+
+    let item (key:'K) =
+        let m = getMap key
+        if isNull m then
+            raise <| KeyNotFoundException(sprintf "Key:%A , does not exist in the dictionary" key)
+        else
+            MapTree.find comparer key m
+    
+    let tryFind (key:'K) =
+        let m = getMap key
+        if isNull m then
+            None
+        else
+            MapTree.tryFind comparer key m
+
+    let tryGetValue (key:'K,out:byref<'V>) =
+        let m = getMap key
+        if isNull m then
+            false
+        else
+            MapTree.tryGetValue comparer key &out m
+
+    let tryFindOpt (key:'K) =
+        let m = getMap key
+        if isNull m then
+            Opt.None
+        else
+            MapTree.tryFindOpt comparer key m
+
+    let resize newBucketSize =
+        let capacityLimit = (newBucketSize * ShardSize) - 1
+        let bmd = calcBitMaskDepth capacityLimit
+        let newBucketBitMask = calcSubBitMask bmd
+        let target = Array.zeroCreate<Shard<'K,'V>> (newBucketSize)
+        //printfn "resizing: capacity:%i bitMaskDepth:%i buckBitMask:%i newBucketSize:%i" capacityLimit bmd newBucketBitMask newBucketSize
+        eachMapInBucket(bucket,fun _ si sm ->
+            MapTree.iter (fun k v -> // for each key in source
+                let kh = k.GetHashCode()
+                let tbi = bucketIndex(kh,newBucketBitMask)
+                let tshrd = getOrCreateShard(target,tbi)
+                let tm = tshrd.[si]
+
+                if isNull tm then
+                    tshrd.[si] <- MapOne(k,v)
+                else                                             
+                    tshrd.[si] <- MapTree.add comparer k v tm
+            ) sm
+        )
+
+
+        for i2 in 0 .. target.Length - 1 do
+            if util. isNull target.[i2] then target.[i2] <- empty
+            
+        //now update internal state
+        lock bucket.SyncRoot (fun () -> bucket <- target ) // ??? needed with resize lock already in place?
+        bitMaskDepth <- bmd
+        bucketBitMask <- newBucketBitMask
+        capacity <- capacityLimit - 1
+
+    let grow () =
+
+        //printfn "started resize ()"
+        let isize = bucket.Length
+        let nsize = isize * 2
+        resize nsize
+        
+            
+    let add(k:'K,v:'V) =
+        // let kh = k.GetHashCode()
+        // let bi = util.bucketIndex(kh,bucketBitMask)
+        // let si = util.shardIndex kh
+        let bi, si = getIndexes(k,bucketBitMask)
```

Comment:
Does the use of a tuple return here cause an unnecessary allocation?

Reply:
I was starting a refactoring and would have preferred the single line allocation helper function, hoping the optimiser may lift this tupling out of the branch but I can revert back to commented out code if this is causing an unneeded allocation (or create a new struct type that has two int fields)